### DISCUSSION TITLE
Securing traffic between namespaces on same machine

### DISCUSSION THREAD

**Discussion by mschrier24** [strongswan/strongswan#1968] [General]
We have successful deployed strongswan in multiple namespaces on a single node, and have connections functioning to external targets.  But often we need applications in different namespaces on the same machine communicate with each other.  This is what we are having problems with - two namespaces on the same node are not able to communicate when there are encrypted policies configured between them.

It might seem obvious that we just wouldn't need secure communication between such namespaces (so why have such connection rules at all, right?).  But our solution is very dynamic - we don't necessarily know which IP address the namespaces will be using, or whether namespaces will be communicating externally across the network or internally within the same node. So we define a set of rules that assumes external network communication, and we expect it to work internally as well.  But unfortunately those rules don't seem to work internally.

We can provide some details, but we initially want to know if this _should_ be possible.  We don't want to have to dynamically change the connection rules based on what is running on a particular namespace, or what state it is running in, since that will cause performance issues.

**Comment by tobiasbrunner** [comment]
I don't see why it shouldn't work in general. But there could obviously be issues with routing or firewall rules.

**Reply by mschrier24** [reply]
OK so one thing I discovered, and I've seen this before.  When there are duplicate policies strongswan sometimes either simply does not handle them correctly, or in some cases it's actually crashed.  So I cleaned things up to provide only namespace-specific policies to the namespace's instance of strongswan, and I am now able to see traffic flow between namespaces.  I'm relying on the bind-mounting feature of netns where I can override ipsec configuration files in /etc.

But I've run into another problem.  When I modify an included configuration file in the /etc/netns path for the namespace strongswan is more often than not taking the version in /etc instead.  For example:

/etc/ipsec.conf --> includes /etc/ipsec.policies
There is also /etc/netns/XXX/ipsec.policies which I want to use for the namespace

The first time I run "ipsec start" in the namespace everything is OK.  When I make a change to the bind-mounted version (in /etc/netns/XXX) and run "ipsec reload" or "ipsec update" on the namespace then strongswan takes the base /etc version of ipsec.policies instead of the namespace version.  I can't understand how it is not taking the bind-mounted version.  Could this be a known bug?  Can you suggest a workaround?  I assume each instance of charon is completely independent and can be updated or loaded independently of all other instances - is this a correct assumption?

**Reply by mschrier24** [reply]
I have learned that there may be some weird behavior with overlay file systems/mounts.  I am modifying both the base file and the overlaid file(s), and I think this is causing problems.  I am going to try keeping the base file fixed, and modify only the overlaid files.  Results to follow...

**Reply by tobiasbrunner** [reply]
> When there are duplicate policies strongswan sometimes either simply does not handle them correctly, or in some cases it's actually crashed.

What do you mean with "duplicate policies"? And crash how/when?

As for the rest, strongSwan has no control over these bind mounts, it simply opens files at the given paths. So whatever issue you have there comes from the tools you use or the system itself.

**Reply by mschrier24** [reply]
To conclude on this, I have re-organized the way the ipsec.conf file includes other files, and am not modifying any "base" files that may interfere with modifications to the corresponding overlaid files.  Base files remain fixed after startup. Things are working as expected now.

**Comment by new-maker** [comment]
When I installed strongswan with the apt tool, I want to run a strongswan instance in each namespace,and then modify the /etc/strongswan.conf text to the following
charon {
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
        piddir = /etc/ipsec.d/run
}
 
include strongswan.d/*.conf

then execute the following commands.what I need to do to run a strongswan instance in each namespace
